### ISSUE TITLE
Add Multi-Label Learning to Rank (MLLTR) feature 

### ISSUE THREAD

**Issue by xueyinya** [microsoft/LightGBM#5363]
## Summary
Our team developed a multi-label learning to rank (MLLTR) framework, which allows us to explore multiple different multi-objective optimization methods to build models that optimizes MLLTR with various preferred balance among different labels / objectives. We would like to open source the MLLTR framework, which is developed based on LightGBM. 

## Motivation
Learning to Rank (LTR) technique is widely used these days, especially in the Search Ranking application. It is desirable to train a model using many relevance criteria instead of just one. Therefore, our team leveraged the Multi-Objective Optimization (MOO) aspect of the MLLTR problem and use the MOO algorithms to solve MLLTR problems. With the MLLTR feature adding into the LightGBM, it would inspire new ideas in this field as well as benefit the related applications.

## Description
Here are some modifications: 
- Basic setups for MLLTR feature, such as config, dataloader, boosting etc. 
- Optimizers and Multi-Gradient combination modified/created to further aid the MLLTR model calculation.
- Modification of discounted cumulative gain (DCG) and related ranking metric as measurements for MLLTR algorithm.
- Extend the feature from tabular data format to libsvm data format.
Note: The user needs to specify several multi-objective (MO) related parameters in the config file to use this feature.

**Comment by jameslamb** [comment]
Thanks very much for your interest in LightGBM!

We're happy to consider this and the related proposals, but I want to be sure to set the right expectation. This project is currently facing a significant lack of maintainer time and attention, and has a large amount of work left ahead of it just to get to the next release of LightGBM's current functionality (#5153).

To start, could you please share links to code so we can understand the scope of what you're proposing adding to LightGBM?

It would also be helpful if you read the related conversation about a proposal to add multi-**task** learning to LightGBM:

* #5047
* #5111

We're going to ask you the same questions that we asked the people proposing that change.

**Comment by jameslamb** [comment]
We'd also really appreciate if you stop creating new issues related to this until we've come to an agreement here about whether or not LightGBM wants to accept MLLTR into this repo at all.

Especially since it seems that those issues are mostly identical copies of this one with one or two sentences changed.

Can you please close #5364, #5365, and #5366 and just include those individual components as bulleted items in the description of this issue?

Then we can use this issue as a central place to discuss whether or not LightGBM will integrate MLLTR at all.

**Comment by xueyinya** [comment]
Hi James,

Thanks for your prompt reply.

I will close the three issues as you mentioned in the previous email. Sorry for any inconvenience.

Best,
Xueying

From: James Lamb ***@***.***>
Reply-To: microsoft/LightGBM ***@***.***>
Date: Monday, July 11, 2022 at 3:08 PM
To: microsoft/LightGBM ***@***.***>
Cc: "Yang, Xueying" ***@***.***>, Author ***@***.***>
Subject: Re: [microsoft/LightGBM] Add Multi-Label Learning to Rank (MLLTR) feature (Part 1: basic setups) (Issue #5363)


We'd also really appreciate if you stop creating new issues related to this until we've come to an agreement here about whether or not LightGBM wants to accept MLLTR into this repo at all.

Especially since it seems that those issues are mostly identical copies of this one with one or two sentences changed.

Can you please close #5364<https://github.com/microsoft/LightGBM/issues/5364>, #5365<https://github.com/microsoft/LightGBM/issues/5365>, and #5366<https://github.com/microsoft/LightGBM/issues/5366> and just include those individual components as bulleted items in the description of this issue?

Then we can use this issue as a central place to discuss whether or not LightGBM will integrate MLLTR at all.

**Comment by jameslamb** [comment]
No problem! Please provide a link to the code you'd like to integrate into LightGBM, so maintainers here can understand the specific details of what you are proposing.
